src/Http/Http/src/Features/FormFeature.cs

```diff
@@ -248,14 +249,17 @@ private async Task<IFormCollection> InnerReadFormAsync(CancellationToken cancell
                         var value = await formDataSection.GetValueAsync();
 
                         formAccumulator.Append(key, value);
-                        if (formAccumulator.ValueCount > _options.ValueCountLimit)
+                        if ((files?.Count ?? 0) + formAccumulator.ValueCount + nonFormOrFileContentDispositionCount > _options.ValueCountLimit)
                         {
                             throw new InvalidDataException($"Form value count limit {_options.ValueCountLimit} exceeded.");
                         }
```

Comment:
I realized this can be much simpler. We only need one `sectionCount`, increment it once on line 194, and then immidiately check against the limit and throw.

Reply:
Updated